File: 0006-zigzag-conversion/0006-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s
        
        res = []
        n = len(s)
        cycleLen = 2 * numRows - 2
        
        for i in range(numRows):
            for j in range(0, n, cycleLen):
                if j + i < n:
                    res.append(s[j + i])
                if i != 0 and i != numRows - 1 and j + cycleLen - i < n:
                    res.append(s[j + cycleLen - i])
        
        return ''.join(res)

File: 0006-zigzag-conversion/0006-zigzag-conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1 || numRows >= s.length()) {
        return s;
    }
    string res;
    int n = s.length();
    int cycleLen = 2 * numRows - 2;
    for (int i = 0; i < numRows; i++) {
        for (int j = 0; j + i < n; j += cycleLen) {
            res += s[j + i];
            if (i != 0 && i != numRows - 1 && j + cycleLen - i < n) {
                res += s[j + cycleLen - i];
            }
        }
    }
    return res;
    }
};